File: src/main/java/com/abysalto/middleware/domain/ProductPage.java
package com.abysalto.middleware.domain;

import java.util.List;

/**
 * A page of products plus the pagination metadata reported by the source.
 *
 * @param items the products on this page
 * @param total total number of products matching the query across all pages
 * @param skip  number of products skipped before this page (offset)
 * @param limit page size requested (0 means "all", per the source contract)
 */
public record ProductPage(List<Product> items, int total, int skip, int limit) {
}

File: dotnet/src/Middleware.Core/Domain/ProductPage.cs
using System.ComponentModel;

namespace Middleware.Core.Domain;

/// <summary>
/// A page of products plus the pagination metadata reported by the source.
/// <para><c>Limit</c> of <c>0</c> means "all", per the source contract (see <see cref="Abstractions.IProductSource"/>).</para>
/// <para>Cached directly by <c>ProductQueryCache</c>, so it carries the same immutability contract as
/// <see cref="Product"/>: <c>Items</c> must be a genuinely immutable collection, not a read-only view
/// over a mutable one.</para>
/// </summary>
[ImmutableObject(true)]
public sealed record ProductPage(IReadOnlyList<Product> Items, int Total, int Skip, int Limit);
